executor/insert_common.go

```diff
@@ -744,11 +744,6 @@ func setDatumAutoIDAndCast(ctx sessionctx.Context, d *types.Datum, id int64, col
 	d.SetAutoID(id, col.GetFlag())
 	var err error
 	*d, err = table.CastValue(ctx, *d, col.ToInfo(), false, false)
-	if err == nil && d.GetInt64() < id {
```

Comment:
Please add some test cases for it.

Reply:
maybe we should add a check about core.Insert.IsReplace in this line 🤔

Reply:
yes, this also works, but we can put `sc := ctx.GetSessionVars().StmtCtx` and `insertPlan, ok := sc.GetPlan().(*core.Insert)` inside the if statement block to avoid more calls.